### PR TITLE
feat. Allow assets to be merged by dest path

### PR DESCRIPTION
**Pull request by juliusl**
- When `merge_assets` is set, will override src/perm by dest path

Address #114

### CODE REVIEW

**Comment by kornelski** [comment]
This exposed a design ambiguity: I've assumed the source file would be the key for the merge (so you can have variant-specific names in the package for the same source file), but you've made the destination the key for merge.

I'm not sure which one is better. Would it make sense to have both behaviors, and merge based on either path? Or maybe it's all too clever and it should just append assets array?

**Comment by juliusl** [comment]
> This exposed a design ambiguity: I've assumed the source file would be the key for the merge (so you can have variant-specific names in the package for the same source file), but you've made the destination the key for merge.
> 
> 
> 
> I'm not sure which one is better. Would it make sense to have both behaviors, and merge based on either path? Or maybe it's all too clever and it should just append assets array?

Interesting, I think adding an option for both behaviors is reasonable. Let me take a look at implementing that.

**Comment by juliusl** [comment]
@kornelski I made changes to allow for merging by dest, src, or by appending. I updated the example with these variants as well. 

Here is a snippet of what usage looks like:

``` 
[package.metadata.deb]
assets = [
    # binary
    ["target/release/example", "usr/bin/", "755"],
    # assets
    ["assets/*", "var/lib/example", "644"],
    ["target/release/assets/*", "var/lib/example", "644"],
    ["3.txt", "var/lib/example/3.txt", "644"],
    ["3.txt", "var/lib/example/merged.txt", "644"],
]

[package.metadata.deb.variants.mergeappend]
merge-assets.append = [
    ["4.txt", "var/lib/example/appended/4.txt", "644"]
]

[package.metadata.deb.variants.mergedest]
merge-assets.by.dest = [
    ["4.txt", "var/lib/example/merged.txt", "644"]
]

[package.metadata.deb.variants.mergesrc]
merge-assets.by.src = [
    ["3.txt", "var/lib/example/merged-2.txt", "644"]
]
```

File: src/manifest.rs [hunk 1]
```diff
@@ -1118,6 +1119,14 @@ impl DependencyList {
     }
 }
 
+/// Type-alias for list of assets
+/// 
+type AssetList = Vec<Vec<String>>;
+
+/// Type-alias for a merge map,
+/// 
+type MergeMap<'a> = BTreeMap<&'a String, [&'a String; 2]>;
```
Comment (kornelski) [inline src/manifest.rs:1128]: These can be &str

File: src/manifest.rs [hunk 2]
```diff
@@ -1152,8 +1162,125 @@ struct CargoDeb {
...
+                        acc
+                    } else {
+                        warn!("Incomplete asset entry {:?}", asset);
+                        acc
+                    }
+                }).iter().map(|(dest, [src, perm])| {
+                    vec![src.to_string(), dest.to_string(), perm.to_string()]
+                }).collect()
+            },
+            MergeByKey::Unknown => todo!(),
```
Comment (kornelski) [inline src/manifest.rs:1263]: Should this be an error? Can you remove the unknown key completely?

**Comment by kornelski** [comment]
Thanks for the pr. This looks more fancy than I expected.

File: src/manifest.rs (hunk 2 above)
Comment (juliusl) [reply src/manifest.rs:1263]: Yep, can definitely remove it.

**Comment by juliusl** [comment]
@kornelski No problem, will you update the docs or should I update them w/ this pr?

**Comment by kornelski** [comment]
Thank you

**Pull request merged**